### PR TITLE
GODRIVER-3339 Separate `Many` and `One` for `UpdateOptions` and `DeleteOptions`.

### PR DESCRIPTION
**Pull request by qingyang-hu**
GODRIVER-3339

## Summary
- Separate the “UpdateOptions” struct into “UpdateManyOptions” and “UpdateOneOptions”.
- Separate the “DeleteOptions” struct into “DeleteManyOptions” and “DeleteOneOptions”.

## Background & Motivation
For [DRIVERS-2822](https://jira.mongodb.org/browse/DRIVERS-2822), `sort` only applies to `UpdateOne`. The server throws an error if a `sort` exists in `UpdateMany`. Therefore, this PR proposes separating the options for `UpdateOne` and `UpdateMany`.
Although there is no immediate requirement yet, the “DeleteOptions” struct is also considered to be divided for future changes as it is shared by both “DeleteOne” and “DeleteMany” now.

### CODE REVIEW

File: internal/integration/unified/crud_helpers.go [hunk 1]
```diff
@@ -20,65 +22,83 @@ func newMissingArgumentError(arg string) error {
...
 	filter bson.Raw
 	update interface{}
-	opts   *options.UpdateOptionsBuilder
+	opts   options.Lister[Options]
 }
 
-func createUpdateArguments(args bson.Raw) (*updateArguments, error) {
-	ua := &updateArguments{
-		opts: options.Update(),
+func createUpdateArguments[Options options.UpdateManyOptions | options.UpdateOneOptions](
```
Comment (prestonvasquez) [inline internal/integration/unified/crud_helpers.go:31]: We should not try to generalize this function. Instead, we should maintain `createUpdateOneArguments` and `createUpdateManyArguments`. Using the combination of reflection and generics here is confusing and difficult to read.

**Review by prestonvasquez — CHANGES_REQUESTED** [review]
We need to update the migration guide to include sections about upgrading from options.Delete() and options.Update() to the new design.

File: mongo/options/updateoptions.go [hunk 2]
```diff
@@ -132,8 +132,144 @@ func (uo *UpdateOptionsBuilder) SetUpsert(b bool) *UpdateOptionsBuilder {
...
+
+	return uo
+}
+
+// UpdateManyOptions represents arguments that can be used to configure UpdateMany
+// operations.
+type UpdateManyOptions struct {
+	// A set of filters specifying to which array elements an update should apply. This option is only valid for MongoDB
+	// versions >= 3.6. For previous server versions, the driver will return an error if this option is used. The
+	// default value is nil, which means the update will apply to all array elements.
```
Comment (matthewdale) [inline mongo/options/updateoptions.go:150]: We don't expect users to interact with the `...Options` structs directly in most cases. We should move the struct field comments to the corresponding setter methods. Leave a comment on each struct field that references the corresponding setter method, like
> See UpdateManyOptionsBuilder.SetArrayFilters for documentation.

This recommendation applies for the new `DeleteManyOptions` also. Note that we will update the pattern for the rest of the options types with [GODRIVER-3327](https://jira.mongodb.org/browse/GODRIVER-3327).

File: mongo/bulk_write.go [hunk 3]
```diff
@@ -423,59 +416,70 @@ func (bw *bulkWrite) runUpdate(ctx context.Context, batch bulkWriteBatch) (opera
...
 	}
 
 	updateDoc = bsoncore.AppendValueElement(updateDoc, "u", u)
 
-	if multi {
-		updateDoc = bsoncore.AppendBooleanElement(updateDoc, "multi", multi)
+	if doc.multi {
+		updateDoc = bsoncore.AppendBooleanElement(updateDoc, "multi", doc.multi)
+	}
+	if doc.sort != nil {
```
Comment (matthewdale) [inline mongo/bulk_write.go:450]: `doc.sort` is currently never set. Will that be used in the follow-up GODRIVER-3285 PRs?

File: internal/integration/unified/crud_helpers.go [hunk 4]
```diff
@@ -63,6 +62,70 @@ func createUpdateArguments(args bson.Raw) (*updateArguments, error) {
...
+	}
+
+	return ua, nil
+}
+
+type updateOneArguments struct {
+	filter bson.Raw
+	update interface{}
+	opts   *options.UpdateOneOptionsBuilder
+}
```
Comment (matthewdale) [inline internal/integration/unified/crud_helpers.go:90]: You can avoid creating a new type here by separating `filter` and `update` from the options builder, allowing you to reuse `updateArguments` in both functions.

E.g. updated function signatures:
```go
func createUpdateOneArguments(args bson.Raw) (*updateArguments, *options.UpdateOneOptionsBuilder, error)
func createUpdateManyArguments(args bson.Raw) (*updateArguments, *options.UpdateManyOptionsBuilder, error)
```

**Review by matthewdale — CHANGES_REQUESTED** [review]

File: mongo/bulk_write.go (hunk 3 above)
Comment (qingyang-hu) [reply mongo/bulk_write.go:450]: You are right. It makes more sense to leave the sort in GODRIVER-3285. I will remove it from this PR.

**Review by matthewdale — APPROVED** [review]
Looks good! 👍

**Review by prestonvasquez — APPROVED** [review]
LGTM!

**Pull request merged**